imap_processing/pointing_frame_handler.py

```diff
@@ -0,0 +1,242 @@
+"""
+Generate Pointing Frame.
+
+Notes
+-----
+Kernels that are required to run this code:
+1. imap_science_0001.tf - pointing frame kernel
+2. imap_sclk_0000.tsc - spacecraft clock kernel
+3. imap_wkcp.tf - spacecraft frame kernel
+4. de430.bsp - standard SPICE planetary ephemeris kernel
+5. naif0012.tls - standard NAIF leapsecond kernel
+6. imap_spin.bc - test attitude kernel available at:
+   https://lasp.colorado.edu/galaxy/display/IMAP/Data
+These need to be placed in tests/pointing_frame/test_data.
+
+References
+----------
+https://spiceypy.readthedocs.io/en/main/documentation.html.
+"""
+
+import logging
+import os
+from pathlib import Path
+
+import numpy as np
+import spiceypy as spice
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+# TODO : Add multiple pointings to the pointing frame.
+
+
+def get_coverage(ck_kernel: str) -> tuple[float, float, np.ndarray]:
+    """
+    Create the pointing frame.
+
+    Parameters
+    ----------
+    ck_kernel : str
+        Path of ck_kernel used to create the pointing frame.
+
+    Returns
+    -------
+    et_start : float
+        Start time of ck_kernel.
+    et_end : float
+        End time of ck_kernel.
+    et_times : numpy.ndarray
+        Array of times between et_start and et_end.
+    """
+    # Each spin is 15 seconds. We want 10 quaternions per spin.
+    # duration / # samples (nominally 15/10 = 1.5 seconds)
+    step = 1.5
+
+    # Get the spacecraft ID.
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.gipool
+    id_imap_spacecraft = spice.gipool("FRAME_IMAP_SPACECRAFT", 0, 1)
+
+    # TODO: Queried pointing start and stop times here.
+
+    # Get the coverage window
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.ckcov
+    cover = spice.ckcov(ck_kernel, int(id_imap_spacecraft), True, "SEGMENT", 0, "TDB")
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.wnfetd
+    et_start, et_end = spice.wnfetd(cover, 0)
+    et_times = np.arange(et_start, et_end, step)
+
+    return et_start, et_end, et_times
+
+
+def average_quaternions(et_times: np.ndarray) -> tuple[np.ndarray, list[np.ndarray]]:
+    """
+    Average the quaternions.
+
+    Parameters
+    ----------
+    et_times : numpy.ndarray
+        Array of times between et_start and et_end.
+
+    Returns
+    -------
+    q_avg : np.ndarray
+        Average quaternion.
+    z_eclip_time : list
+        Z-axis of the ECLIPJ2000 frame. Used for plotting.
+    """
+    z_eclip_time = []
+    aggregate = np.zeros((4, 4))
+
+    for tdb in et_times:
```

Comment:
I would like to keep Nick's comment in here about this method. Personally, I find it odd that we are not getting the quaternions directly from the CK, but instead we are resampling.
```suggestion
    for tdb in et_times:
        # we use a quick and dirty method here for grabbing the quaternions
        # from the attitude kernel.  Depending on how well the kernel input
        # data is built and sampled, there may or may not be aliasing with this
        # approach.  If it turns out that we need to pull the quaternions
        # direclty from the CK there are several routines that exist to do this
        # but it's not straight forward.  We'll revisit this if needed.
        #
        # CKLPF - open the ck
        # DAFBBS - begin backwards order search for segments
        # DAFGS - gets the segment summary
        # CKNR03 - gets the number of records in a type 3 CK segment
        # CKGR03 - retrieves an individual record from a type 3 CK segment
        # CKUPF - close the file
        #
```

Reply:
Sounds good.